tests/integration/Core/Repository/URLWildcardService/CriterionTest.php

```diff
@@ -0,0 +1,290 @@
+            '/ibexa-dxp',
+            '/facebook',
+        ];
+
+        $query = new URLWildcardQuery();
+        $query->filter = new Criterion\Type(true);
+
+        $searchResult = $this->findUrlWildcards($query, count($expectedWildcardUrls));
+        $this->checkWildcardUrl($searchResult->items, $expectedWildcardUrls);
+    }
+
+    public function testTypeNoForward(): void
+    {
+        $expectedWildcardUrls = [
+            '/nice-url-seo',
+            '/no-forward test url',
+            '/Twitter',
+        ];
+
+        $query = new URLWildcardQuery();
+        $query->filter = new Criterion\Type(false);
+
+        $searchResult = $this->findUrlWildcards($query, count($expectedWildcardUrls));
+        $this->checkWildcardUrl($searchResult->items, $expectedWildcardUrls);
+    }
+
+    public function testInvalidLimitThrowsInvalidArgumentException()
+    {
+        $query = new URLWildcardQuery();
+        $query->filter = new Criterion\MatchAll();
+        $query->limit = 'invalid!';
+
+        $repository = $this->getRepository();
+        $urlWildcardService = $repository->getURLWildcardService();
+
+        $this->expectException(InvalidArgumentValue::class);
+        $urlWildcardService->findUrlWildcards($query);
+    }
+
+    public function testInvalidOffsetThrowsInvalidArgumentException()
+    {
+        $query = new URLWildcardQuery();
+        $query->filter = new Criterion\MatchAll();
+        $query->offset = 'invalid!';
+
+        $repository = $this->getRepository();
+        $urlWildcardService = $repository->getURLWildcardService();
+
+        $this->expectException(InvalidArgumentValue::class);
+        $urlWildcardService->findUrlWildcards($query);
+    }
+
+    public function testSourceAndDestination(): void
+    {
+        $search = 'test';
+        $expectedWildcardUrlsSource = [
+            '/test',
+            '/test test',
+            '/no-forward test url',
+        ];
+
+        $expectedWildcardUrlsDestination = [
+            '/content-test',
+            '/content test',
+            '/no/forward test url',
+        ];
+
+        $query = new URLWildcardQuery();
+        $query->filter = new Criterion\LogicalAnd([
+            new Criterion\SourceUrl($search),
+            new Criterion\DestinationUrl($search),
+        ]);
+
+        $searchResult = $this->findUrlWildcards($query, count($expectedWildcardUrlsSource));
+
+        $this->checkWildcardUrl($searchResult->items, $expectedWildcardUrlsSource);
+        $this->checkWildcardUrl($searchResult->items, $expectedWildcardUrlsDestination, false);
+    }
+
+    public function testLogicalInvalidCriterion(): void
+    {
+        $this->expectException(InvalidCriterionArgumentException::class);
+        $this->expectExceptionMessage("You provided eZ\Publish\API\Repository\Values\URL\Query\Criterion\VisibleOnly at index '1', but only instances of 'Ibexa\Contracts\Core\Repository\Values\Content\URLWildcard\Query\Criterion' are accepted");
```

Comment:
I'd go for readability here, this line is rather too long.
```suggestion
        $this->expectExceptionMessage(
            'You provided eZ\Publish\API\Repository\Values\URL\Query\Criterion\VisibleOnly ' .
            "at index '1', but only instances of " .
            "'Ibexa\Contracts\Core\Repository\Values\Content\URLWildcard\Query\Criterion' are accepted"
        );
```